src/compiler/scala/tools/nsc/backend/jvm/opt/BoxUnbox.scala

```diff
@@ -326,31 +329,38 @@ abstract class BoxUnbox {
           if (boxKind.boxedTypes.lengthCompare(1) == 0) {
             // fast path for single-value boxes
             allConsumers.foreach(extraction => extraction.postExtractionAdaptationOps(boxKind.boxedTypes.head) match {
-              case Nil =>
-                toDelete ++= extraction.allInsns
+              case Nil => extraction match {
+                case Drop(_) => toReplace(extraction.consumer) = boxKind.boxedTypes.map(t => getPop(t.getSize))
+                case _ => toDelete ++= extraction.allInsns
+              }
               case ops =>
                 toReplace(extraction.consumer) = ops
                 toDelete ++= extraction.allInsns - extraction.consumer
             })
           } else {
             for (extraction <- allConsumers) {
-              val valueIndex = boxKind.extractedValueIndex(extraction)
-              val replacementOps = if (valueIndex == 0) {
-                val pops = boxKind.boxedTypes.tail.map(t => getPop(t.getSize))
-                pops ::: extraction.postExtractionAdaptationOps(boxKind.boxedTypes.head)
-              } else {
-                var loadOps: List[AbstractInsnNode] = null
-                val consumeStack = boxKind.boxedTypes.zipWithIndex reverseMap {
-                  case (tp, i) =>
-                    if (i == valueIndex) {
-                      val resultSlot = getLocal(tp.getSize)
-                      loadOps = new VarInsnNode(tp.getOpcode(ILOAD), resultSlot) :: extraction.postExtractionAdaptationOps(tp)
-                      new VarInsnNode(tp.getOpcode(ISTORE), resultSlot)
-                    } else {
-                      getPop(tp.getSize)
+              val replacementOps = extraction match {
+                case Drop(_) =>
+                  boxKind.boxedTypes.map(t => getPop(t.getSize))
```

Comment:
This should be `reverseMap`.

```
package p1

class Test {

  def crash(progress: Int): Unit = {
    val limit = 5
    var n = 1
    var (dormancy, factor) = progress match {
      case 0 => (10000L, 5)
      case 1 => (250L, 4)
    }
    ()
  }
}


```

```
Caused by: scala.tools.asm.tree.analysis.AnalyzerException: Illegal use of POP2
	at scala.tools.asm.tree.analysis.Frame.execute(Frame.java:366)
	at scala.tools.nsc.backend.jvm.analysis.AliasingFrame.execute(AliasingFrame.scala:113)
	at scala.tools.asm.tree.analysis.Analyzer.analyze(Analyzer.java:188)
	... 63 more
error: Error while emitting p1/Test
While processing p1/Test.crash
one error found
```

The other test case I minimized in https://github.com/scala/scala/pull/9495 doesn't crash in 2.12.x but it might be due to good luck.